third-party/cmake/BuildLuarocks.cmake

```diff
@@ -1,35 +1,83 @@
-    -DURL=${LUAROCKS_URL}
-    -DEXPECTED_SHA256=${LUAROCKS_SHA256}
-    -DTARGET=luarocks
-    -P ${CMAKE_CURRENT_SOURCE_DIR}/cmake/DownloadAndExtractFile.cmake
-  BUILD_IN_SOURCE 1
-  CONFIGURE_COMMAND ${DEPS_BUILD_DIR}/src/luarocks/configure
-    --prefix=${DEPS_INSTALL_DIR} --force-config ${LUAROCKS_OPTS}
-    --lua-suffix=jit
-  BUILD_COMMAND ""
-  INSTALL_COMMAND ${MAKE_PRG} bootstrap)
+# Reusable function to build Luarocks, set the following variables
+# before calling:
+#  LUAROCKS_CONFIGURE_COMMAND
+#  LUAROCKS_BUILD_COMMAND
+#  LUAROCKS_INSTALL_COMMAND
+function(BuildLuarocks)
+  ExternalProject_Add(luarocks
+    PREFIX ${DEPS_BUILD_DIR}
+    URL ${LUAROCKS_URL}
+    DOWNLOAD_DIR ${DEPS_DOWNLOAD_DIR}/luarocks
+    DOWNLOAD_COMMAND ${CMAKE_COMMAND}
+      -DPREFIX=${DEPS_BUILD_DIR}
+      -DDOWNLOAD_DIR=${DEPS_DOWNLOAD_DIR}/luarocks
+      -DURL=${LUAROCKS_URL}
+      -DEXPECTED_SHA256=${LUAROCKS_SHA256}
+      -DTARGET=luarocks
+      -P ${CMAKE_CURRENT_SOURCE_DIR}/cmake/DownloadAndExtractFile.cmake
+    BUILD_IN_SOURCE 1
+    CONFIGURE_COMMAND "${LUAROCKS_CONFIGURE_COMMAND}"
+    BUILD_COMMAND "${LUAROCKS_BUILD_COMMAND}"
+    INSTALL_COMMAND "${LUAROCKS_INSTALL_COMMAND}")
+endfunction()
 
-list(APPEND THIRD_PARTY_DEPS luarocks)
+# The luarocks binary location
+set(LUAROCKS_BINARY ${HOSTDEPS_BIN_DIR}/luarocks)
+
+# Arguments for calls to 'luarocks build'
+if(MSVC)
+  # In native Win32 don't pass the compiler/linker to luarocks, the bundled
+  # version already knows, and passing them here breaks the build
+  set(LUAROCKS_BUILDARGS CFLAGS=/MT)
+else()
+  set(LUAROCKS_BUILDARGS CC=${HOSTDEPS_C_COMPILER} LD=${HOSTDEPS_C_COMPILER})
+endif()
+
+if(UNIX OR (MINGW AND CMAKE_CROSSCOMPILING))
+
+  if(USE_BUNDLED_LUAJIT)
+    list(APPEND LUAROCKS_OPTS
+      --with-lua=${HOSTDEPS_INSTALL_DIR}
+      --with-lua-include=${HOSTDEPS_INSTALL_DIR}/include/luajit-2.0)
+  endif()
 
-# The path to the luarocks executable
-set(LUAROCKS_BINARY ${DEPS_BIN_DIR}/luarocks)
-# Common build arguments for luarocks build
-set(LUAROCKS_BUILDARGS CC=${DEPS_C_COMPILER} LD=${DEPS_C_COMPILER})
+  set(LUAROCKS_CONFIGURE_COMMAND ${DEPS_BUILD_DIR}/src/luarocks/configure
+      --prefix=${HOSTDEPS_INSTALL_DIR} --force-config ${LUAROCKS_OPTS}
+      --lua-suffix=jit)
+  set(LUAROCKS_BUILD_COMMAND "")
+  set(LUAROCKS_INSTALL_COMMAND ${MAKE_PRG} bootstrap)
+  BuildLuarocks()
+elseif(MSVC)
+
+  # Ignore USE_BUNDLED_LUAJIT - always ON for native Win32
+  set(LUAROCKS_CONFIGURE_COMMAND "")
+  set(LUAROCKS_BUILD_COMMAND "")
+  set(LUAROCKS_INSTALL_COMMAND install.bat /FORCECONFIG /NOREG /NOADMIN /Q /F
+    /LUA ${DEPS_INSTALL_DIR}
+    /LIB ${DEPS_LIB_DIR}
+    /BIN ${DEPS_BIN_DIR}
+    /INC ${DEPS_INSTALL_DIR}/include/luajit-2.0/
+    /P ${DEPS_INSTALL_DIR} /TREE ${DEPS_INSTALL_DIR}
+    /SCRIPTS ${DEPS_BIN_DIR}
+    /CMOD ${DEPS_BIN_DIR}
+    /LUAMOD ${DEPS_BIN_DIR}/lua)
+  BuildLuarocks()
+  set(LUAROCKS_BINARY ${DEPS_INSTALL_DIR}\\2.2\\luarocks.bat)
```

Comment:
I think so too. I don't have a native windows machine ATM, push a commit to see how Appveyor feels about it.